pkg/pgmodel/Readme.md

```diff
@@ -36,46 +80,51 @@ type Sample {
+2. Groups the `model.Series` together with its `[]Samples` as a `model.Insertable`.
+3. Puts all `model.Insertable` entries into a map, keyed by the metric name (`map[string][]model.Insertable`)
 
-After a batch is finished by the metric layer its forwarded to an arbitrary physical inserter, which translates the insert request into the actual SQL doing the insert, sends the request to the DB, and reports the result back to the original goroutine. We maintain a fixed number of physical inserters based on the number of connections to the DB we possess.
+The ingestor passes this map on to the dispatcher (`dispatcher.InsertTs(...)`).
 
-Diagrammatically the write path cn be visualized as follows:
+#### Dispatcher
 
-```
-|, /, \ are go channels
-= is for network connections
-- is for normal function calls
+The dispatcher processes this map by iterating through the `(key, value)` pairs (which are metric name, and the insertable (`model.Insertable`)), and sends it on a channel to the one metric batcher goroutine responsible for batching samples for that metric. If the metric batcher does not exist yet, it is created on demand.
 
-                  new req
-                     v
-                 +--------+        +-------------+
-                 | Ingest | <----> | Label Cache |
-                 +--------+        +-------------+
-                  /     \
-                 /       \
- +----------------+     +----------------+       +-----------------+
- | metric handler | ... | metric handler | <===> | DB (series IDs) |
- +----------------+     +----------------+       +-----------------+
-                 \       /
-                  \     /
-                   \   /
-                    \ /
- +----------------+  |  +----------------+       +------------------+
- | insert handler | ... | insert handler | ====> | DB (metric data) |
- +----------------+     +----------------+       +------------------+
-```
+If asyncAcks=true, the dispatcher returns immediately, otherwise it waits for the metric batcher goroutine to return a result, and returns that in response.
+
+### Metric batcher
+
+The second stage consists of the metric batcher, one goroutine for each metric name. These goroutines are created in `dispatcher.InsertTs`, but are pooled for later re-use.
+
+On the first new insert request, the metric batcher creates a metric table in the database, if it doesn't exist yet.
+
+From then onwards, when the metric batcher receives a new insert request on its channel it: 
+1. Creates an unbuffered data-transfer channel for this batcher
+2. Initiates a read request to the copier on the copier read-request channel, sending a handle to the batcher's data-transfer channel
+3. Continues to (in a loop over `select`):
+   1. accept data for the metric (from the dispatcher) on its channel, writing them into a `model.Batch`, until its maximum batch size is reached (`metrics.FlushSize`, currently 2000)
```

Comment:
It will even try sending a batch before reaching the max size if possible (if copier channel is not full), otherwise it will keep batching until max size and block until it sends the batch.

Reply:
done